File: polyn_messages/lib/naming.ex
defmodule Polyn.Naming do
  @moduledoc """
  Utilities for working with Polyn naming conventions
  """

  @doc """
  Validate the name of a message

  ## Examples

      iex>Polyn.Naming.validate_message_name("user.created")
      :ok

      iex>Polyn.Naming.validate_message_name("user  created")
      {:error, message}
  """
  @spec validate_message_name(name :: binary()) :: :ok | {:error, binary()}
  def validate_message_name(name) do
    if String.match?(name, ~r/^[a-z0-9]+(?:\.[a-z0-9]+)*$/) do
      :ok
    else
      {:error, "Message names must be lowercase, alphanumeric and dot separated"}
    end
  end

  @doc """
  Validate the name of a message. Raises if invalid

  ## Examples

      iex>Polyn.Naming.validate_message_name!("user.created")
      :ok

      iex>Polyn.Naming.validate_message_name!("user  created")
      Polyn.ValidationException
  """
  @spec validate_message_name!(name :: binary()) :: :ok
  def validate_message_name!(type) do
    case validate_message_name(type) do
      {:error, reason} ->
        raise Polyn.NamingException, reason

      success ->
        success
    end
  end
end

File: polyn_cli/lib/polyn/cli/naming.rb
# frozen_string_literal: true

module Polyn
  class Cli
    class Naming
      ##
      # Convert a dot separated name into a colon separated name
      def self.dot_to_colon(str)
        str.gsub(".", ":")
      end

      def self.validate_stream_name!(name)
        return if name.match(/^[a-zA-Z0-9_]+$/)

        raise Polyn::Cli::Error,
          "Stream name must be all alphanumeric, uppercase, and underscore separated. Got #{name}"
      end

      def self.format_stream_name(name)
        name.upcase
      end

      def self.validate_destination_name!(name)
        return if name.is_a?(String) && name.match?(/\A[a-z0-9]+(?:(?:\.|:)[a-z0-9]+)*\z/)

        raise Polyn::Cli::Error,
          "Message destination must be lowercase, alphanumeric and dot/colon separated, got #{name}"
      end

      def self.validate_message_name!(name)
        return if name.is_a?(String) && name.match?(/\A[a-z0-9]+(?:\.[a-z0-9]+)*\z/)

        raise Polyn::Cli::Error,
          "Message names must be lowercase, alphanumeric and dot separated"
      end

      def self.dot_to_underscore(name)
        name.gsub(".", "_")
      end

      def self.colon_to_underscore(name)
        name.gsub(":", "_")
      end
    end
  end
end
